flink-runtime/src/test/java/org/apache/flink/runtime/operators/util/OutputEmitterTest.java

```diff
@@ -48,404 +49,272 @@
 		} catch (DeserializationException re) {
 			return;
 		}
 		Assert.fail("Expected a NullKeyFieldException.");
 	}
-	
-	@SuppressWarnings({"serial", "rawtypes"})
+
+	private boolean verifyWrongPartitionHashKey(int position, int fieldNum) {
+		final TypeComparator<Record> comparator = new RecordComparatorFactory(
+			new int[] {position}, new Class[] {IntValue.class}).createComparator();
+		final ChannelSelector<SerializationDelegate<Record>> selector = new OutputEmitter<>(
+			ShipStrategyType.PARTITION_HASH, comparator);
+		final SerializationDelegate<Record> delegate = new SerializationDelegate<>(new RecordSerializerFactory().getSerializer());
+
+		Record record = new Record(2);
+		record.setField(fieldNum, new IntValue(1));
+		delegate.setInstance(record);
+
+		try {
+			selector.selectChannels(delegate, 100);
+		} catch (NullKeyFieldException re) {
+			Assert.assertEquals(position, re.getFieldNumber());
+			return true;
+		}
+		return false;
+	}
+
+	private int[] getSelectedChannels(
+		ShipStrategyType shipStrategyType,
+		int numRecords,
+		int numChannels,
+		Either<Integer, String> recordType) {
+		final TypeComparator<Record> comparator = new RecordComparatorFactory(
+			new int[] {0}, new Class[] {recordType.isLeft() ? IntValue.class : StringValue.class}).createComparator();
+		final ChannelSelector<SerializationDelegate<Record>> selector = new OutputEmitter<>(shipStrategyType, comparator);
+		final SerializationDelegate<Record> delegate = new SerializationDelegate<>(new RecordSerializerFactory().getSerializer());
+		return getSelectedChannels(selector, delegate, recordType, numRecords, numChannels);
+	}
+
+	private int[] getSelectedChannels(
+		ChannelSelector<SerializationDelegate<Record>> selector,
+		SerializationDelegate<Record> delegate,
+		Either<Integer, String> recordType,
+		int numRecords,
+		int numChannels) {
+		int[] hits = new int[numChannels];
+		Value value;
+		for (int i = 0; i < numRecords; i++) {
+			if (recordType.isLeft()) {
+				value = new IntValue(i);
+			} else {
+				value = new StringValue(i + "");
+			}
+			Record record = new Record(value);
+			delegate.setInstance(record);
+
+			int[] channels = selector.selectChannels(delegate, hits.length);
+			for (int channel : channels) {
+				hits[channel]++;
+			}
+		}
+		return hits;
+	}
+
+	private void assertPartitionHashSelectedChannels(
+		ChannelSelector selector,
+		SerializationDelegate<Integer> serializationDelegate,
+		int record,
+		int numChannels) {
+		serializationDelegate.setInstance(record);
+		int[] selectedChannels = selector.selectChannels(serializationDelegate, numChannels);
+
+		assertTrue(selectedChannels.length == 1);
+		assertTrue(selectedChannels[0] >= 0 && selectedChannels[0] <= numChannels - 1);
+	}
+
+	private void verifyPartitionHashSelectedChannels(int numRecords, int numChannels, Either<Integer, String> recordType) {
+		int[] hits = getSelectedChannels(ShipStrategyType.PARTITION_HASH, numRecords, numChannels, recordType);
```

Comment:
nit: maybe rename `getSelectedChannels` to `getSelectedChannelsHitCount`? It's indeed difficult to name this method, but `int[] hits = getSelectedChannels()` looks strange and I had to go through the implementation to understand what is happening :)